File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleSettingRow.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.minimumInteractiveComponentSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.foundation.BorderStroke
import androidx.compose.material3.HorizontalDivider
import androidx.compose.ui.unit.dp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** A settings row: a glyph tile, a title with an optional supporting line, and a trailing control. */
@Composable
fun UseSmileIDSampleSettingRow(
    title: String,
    modifier: Modifier = Modifier,
    supportingText: String? = null,
    testId: String? = null,
    onClick: (() -> Unit)? = null,
    leading: @Composable ((Color) -> Unit)? = null,
    trailing: @Composable (() -> Unit)? = null,
) {
    val colors = UseSmileIDSampleTheme.colors
    Row(
        modifier = modifier
            .fillMaxWidth()
            .then(if (onClick != null) Modifier.clickable(role = Role.Button, onClick = onClick) else Modifier)
            .defaultMinSize(minHeight = SmileDimens.space64)
            .padding(horizontal = SmileDimens.spacingMd, vertical = SmileDimens.spacingSm)
            .tagged(testId),
        horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (leading != null) {
            Surface(
                    modifier = Modifier.size(TILE_SIZE),
                shape = RoundedCornerShape(TILE_RADIUS),
                color = colors.surfaceTile,
            ) {
                Box(contentAlignment = Alignment.Center) { leading(colors.textTitle) }
            }
        }
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXxs),
        ) {
            Text(
                text = title,
                style = UseSmileIDSampleTheme.type.textStyleBodyStrong,
                color = colors.textTitle,
            )
            if (supportingText != null) {
                Text(
                    text = supportingText,
                    style = UseSmileIDSampleTheme.type.textStyleCaption,
                    color = colors.textMuted,
                )
            }
        }
        trailing?.invoke()
    }
}

/** The trailing chevron that says the row pushes a screen. */
@Composable
fun UseSmileIDSampleSettingRowChevron() =
    ChevronRightGlyph(tint = UseSmileIDSampleTheme.colors.textMuted, size = CHEVRON_SIZE)

/** The rule between rows inside one section card. */
@Composable
fun UseSmileIDSampleSettingRowDivider() = HorizontalDivider(
    thickness = SmileDimens.borderWidthHairline,
    color = UseSmileIDSampleTheme.colors.card.border,
)

/** Sign out: full width, centred, error-coloured. */
@Composable
fun UseSmileIDSampleDestructiveRow(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    testId: String? = null,
) {
    Surface(
        onClick = onClick,
        modifier = modifier
            .fillMaxWidth()
            .minimumInteractiveComponentSize()
            .tagged(testId),
        shape = RoundedCornerShape(SmileDimens.radiusSurface),
        color = UseSmileIDSampleTheme.colors.surface,
        border = BorderStroke(SmileDimens.borderWidthHairline, UseSmileIDSampleTheme.colors.card.border),
    ) {
        Box(
            modifier = Modifier.padding(SmileDimens.spacingSm),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                // The soft error text, at the button size — errorFill is the saturated pill colour.
                text = text,
                style = UseSmileIDSampleTheme.type.textStyleButton,
                color = UseSmileIDSampleTheme.colors.badge.errorText,
                textAlign = TextAlign.Center,
            )
        }
    }
}

private val TILE_SIZE = 38.dp
private val TILE_RADIUS = 11.dp
private val CHEVRON_SIZE = 14.dp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_setting_row.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_product_hues.dart';
import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_glyphs.dart';

/// A settings row: a glyph tile, a title with an optional supporting line, and a trailing control.
class UseSmileIDSampleSettingRow extends StatelessWidget {
  /// [leading] is handed the tint the tile resolves to, so no caller picks a colour.
  const UseSmileIDSampleSettingRow({
    required this.title,
    this.supportingText,
    this.onTap,
    this.leading,
    this.trailing,
    this.testId,
    super.key,
  });

  /// The row's title.
  final String title;

  /// The line beneath it, where the design draws one.
  final String? supportingText;

  /// What a tap does; absent on a row whose only control is its trailing one.
  final VoidCallback? onTap;

  /// The row's own mark, which the design supplies for all eleven rows.
  final Widget Function(Color tint)? leading;

  /// The switch or chevron at the end of the row.
  final Widget? trailing;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    final Widget row = ConstrainedBox(
      constraints: const BoxConstraints(minHeight: SmileDimens.space64),
      child: Padding(
        padding: const EdgeInsets.symmetric(
          horizontal: SmileDimens.spacingMd,
          vertical: SmileDimens.spacingSm,
        ),
        child: Row(
          children: <Widget>[
            if (leading != null) ...<Widget>[
              Container(
                width: _tileSize,
                height: _tileSize,
                alignment: Alignment.center,
                decoration: BoxDecoration(
                  // surface-2, a cool grey — surface-alt is a warm cream and drew peach tiles.
                  color: colors.surfaceTile,
                  borderRadius: BorderRadius.circular(_tileRadius),
                ),
                child: leading!(colors.textTitle),
              ),
              const SizedBox(width: SmileDimens.spacingSm),
            ],
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                mainAxisSize: MainAxisSize.min,
                children: <Widget>[
                  Text(
                    title,
                    style: UseSmileIDSampleType.textStyleBodyStrong.copyWith(
                      color: colors.textTitle,
                    ),
                  ),
                  if (supportingText != null) ...<Widget>[
                    const SizedBox(height: SmileDimens.spacingXxs),
                    Text(
                      supportingText!,
                      style: UseSmileIDSampleType.textStyleCaption.copyWith(
                        color: colors.textMuted,
                      ),
                    ),
                  ],
                ],
              ),
            ),
            if (trailing != null) ...<Widget>[
              const SizedBox(width: SmileDimens.spacingSm),
              trailing!,
            ],
          ],
        ),
      ),
    );
    return Semantics(
      identifier: testId,
      button: onTap != null,
      child: onTap == null
          ? row
          : Material(
              color: Colors.transparent,
              child: InkWell(onTap: onTap, child: row),
            ),
    );
  }
}

/// The trailing chevron that says the row pushes a screen.
class UseSmileIDSampleSettingRowChevron extends StatelessWidget {
  /// Takes nothing; its size and tint are the design's.
  const UseSmileIDSampleSettingRowChevron({super.key});

  @override
  Widget build(BuildContext context) => UseSmileIDSampleGlyphs.chevronRight(
    UseSmileIDSampleTheme.colorsOf(context).textMuted,
    size: _chevronSize,
  );
}

/// The rule between rows inside one section card, on the same pair as the card's own outline.
class UseSmileIDSampleSettingRowDivider extends StatelessWidget {
  /// Takes nothing; the pair it draws is the one the card is outlined in.
  const UseSmileIDSampleSettingRowDivider({super.key});

  @override
  Widget build(BuildContext context) => Divider(
    height: smileCardStrokeWidth,
    thickness: smileCardStrokeWidth,
    color: UseSmileIDSampleTheme.colorsOf(context).cardStroke,
  );
}

/// Sign out: full width, centred, in the soft error text rather than the saturated fill.
class UseSmileIDSampleDestructiveRow extends StatelessWidget {
  /// Takes the copy, because "Sign out" is product text rather than a component's.
  const UseSmileIDSampleDestructiveRow({
    required this.text,
    required this.onTap,
    this.testId,
    super.key,
  });

  /// The action's label.
  final String text;

  /// What a tap does.
  final VoidCallback onTap;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Semantics(
      identifier: testId,
      button: true,
      child: Material(
        color: colors.surface,
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(SmileDimens.radiusSurface),
          side: BorderSide(
            color: colors.cardStroke,
            width: smileCardStrokeWidth,
          ),
        ),
        child: InkWell(
          onTap: onTap,
          borderRadius: BorderRadius.circular(SmileDimens.radiusSurface),
          child: ConstrainedBox(
            constraints: const BoxConstraints(
              minHeight: SmileDimens.sizeControlMd,
            ),
            child: Padding(
              padding: const EdgeInsets.all(SmileDimens.spacingSm),
              child: Center(
                child: Text(
                  text,
                  textAlign: TextAlign.center,
                  // The soft error text; errorFill is the saturated pill colour.
                  style: UseSmileIDSampleType.textStyleButton.copyWith(
                    color: colors.badge.errorText,
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

/// The design's tile, which no scale token carries.
const double _tileSize = 38;
const double _tileRadius = 11;

/// The design's chevron, smaller than the icon scale's 16.
const double _chevronSize = 14;
